genkit-tools/common/src/eval/localFileDatasetStore.ts

```diff
@@ -0,0 +1,221 @@
+ * limitations under the License.
+ */
+
+import crypto from 'crypto';
+import fs from 'fs';
+import { appendFile, readFile, rm, writeFile } from 'fs/promises';
+import os from 'os';
+import path from 'path';
+import { v4 as uuidv4 } from 'uuid';
+import { CreateDatasetRequest, UpdateDatasetRequest } from '../types';
+import {
+  Dataset,
+  DatasetMetadata,
+  DatasetMetadataSchema,
+  DatasetStore,
+  EvalFlowInputSchema,
+} from '../types/eval';
+import { logger } from '../utils/logger';
+
+/**
+ * A local, file-based DatasetStore implementation.
+ */
+export class LocalFileDatasetStore implements DatasetStore {
+  private readonly storeRoot;
+  private readonly indexFile;
+  private readonly INDEX_DELIMITER = '\n';
+  private static cachedDatasetStore: LocalFileDatasetStore | null = null;
+
+  private constructor() {
+    this.storeRoot = this.generateRootPath();
+    this.indexFile = this.getIndexFilePath();
+    fs.mkdirSync(this.storeRoot, { recursive: true });
+    if (!fs.existsSync(this.indexFile)) {
+      fs.writeFileSync(path.resolve(this.indexFile), '');
+    }
+    logger.info(
+      `Initialized local file dataset store at root: ${this.storeRoot}`
+    );
+  }
+
+  static getDatasetStore() {
+    if (!this.cachedDatasetStore) {
+      this.cachedDatasetStore = new LocalFileDatasetStore();
+    }
+    return this.cachedDatasetStore;
+  }
+
+  static reset() {
+    this.cachedDatasetStore = null;
+  }
+
+  async createDataset(req: CreateDatasetRequest): Promise<DatasetMetadata> {
+    return this.createDatasetInternal(req.data, req.displayName);
+  }
+
+  private async createDatasetInternal(
+    data: Dataset,
+    displayName?: string
+  ): Promise<DatasetMetadata> {
+    const datasetId = this.generateDatasetId();
+    const filePath = path.resolve(
+      this.storeRoot,
+      this.generateFileName(datasetId)
+    );
+
+    if (!fs.existsSync(filePath)) {
+      logger.error(`Dataset already exists at ` + filePath);
+      throw new Error(
+        `Create dataset failed: file already exists at {$filePath}`
+      );
+    }
+
+    logger.info(`Saving Dataset to ` + filePath);
+    await writeFile(filePath, JSON.stringify(data));
+
+    const metadata = {
+      datasetId: datasetId,
+      size: Array.isArray(data) ? data.length : data.samples.length,
+      uri: 'file://' + filePath,
+      version: 1,
+      displayName: displayName,
+      createTime: Date.now().toString(),
+      updateTime: Date.now().toString(),
+    };
+
+    logger.debug(
+      `Saving DatasetMetadata for ID ${datasetId} to ` +
+        path.resolve(this.indexFile)
+    );
+    await appendFile(
+      path.resolve(this.indexFile),
+      JSON.stringify(metadata) + this.INDEX_DELIMITER
+    );
+    return metadata;
+  }
+
+  async updateDataset(req: UpdateDatasetRequest): Promise<DatasetMetadata> {
+    if (!req.datasetId) {
+      return this.createDatasetInternal(req.patch, req.displayName);
+    }
```

Comment:
Should we support this behavior?

Reply:
Left it as is for now. Asked on chat, please take another look and let me know.

Reply:
Just throw if datasetId is missing.

Reply:
Removed since we don't need this after making it required.